### PR TITLE
simplification of the binary tree with the distributive property

### PR DESCRIPTION
**Pull request by universenox**

### CODE REVIEW

**Review by Laouen — CHANGES_REQUESTED** [review]
It seems you should have first merged or rebased master before making the PR to fix conflicts

**Comment by universenox** [comment]
Okay, I fixed it.

File: include/real/irrational_helpers.hpp [hunk 1]
```diff
@@ -3,10 +3,14 @@
 
 #include <vector>
 #include <algorithm>
+#include <real/real.hpp>
+#include<math.h>
```
Comment (sdavtaker) [inline include/real/irrational_helpers.hpp]: White space consistency.

File: include/real/irrational_helpers.hpp [hunk 2]
```diff
@@ -46,6 +50,56 @@ namespace boost {
...
+                    k = k + one_n;
+                }
+
+                return current_value.digits[n];
+            }
+
+            int pi_get_nth_digit(unsigned int n) {
+                if (n > 2000)
+                    throw(pi_precision_exception());
+                // 1000 digits of pi
```
Comment (sdavtaker) [inline include/real/irrational_helpers.hpp]: You throwing when asking for more than 2000 digits 2 lines above but providing only 1000. Why?

File: include/real/real.hpp [hunk 3]
```diff
@@ -247,14 +431,13 @@ namespace boost {
              *
              * @param other - the right side operand boost::real::real number.
              */
-            void operator+=(real& other) {
-                // do not want to overwrite *this->_real_p if others are pointing to it
-                // if others are pointing to it, point to a copy in newly allocated memory, then create operation 
-                if(this->_real_p.use_count() > 1) {
-                    this->_real_p = std::make_shared<real_data>(real_data(*this->_real_p));
+            void operator+=(real other) {
+                auto p = check_and_distribute(other, true, OPERATION::ADDITION);
```
Comment (sdavtaker) [inline include/real/real.hpp]: I would use structured binding here:
bool b, auto p = check...

File: include/real/exact_number.hpp [hunk 4]
```diff
@@ -4,14 +4,20 @@
 #include <vector>
 #include <string>
 #include <algorithm>
+#include <math.h>
+#include <type_traits>
 
 namespace boost {
     namespace real {
         struct exact_number {
+            typedef int exponent_t;
```
Comment (sdavtaker) [inline include/real/exact_number.hpp]: please use "using syntax" for typedefs.
using exponent_t=int;

File: include/real/exact_number.hpp [hunk 5]
```diff
@@ -171,6 +177,242 @@ namespace boost {
...
+                }
+
+                if ((index == 0) && keep_carrying) { // i.e., .999 should become 1.000
+                    if(this->digits[index] == DIGIT_MAX) {
+                        this->digits[index] = 0;
+                        this->push_front(1);
+                        ++this->exponent;
+                    }
+                    else
+                        ++this->digits[index];
```
Comment (sdavtaker) [inline include/real/exact_number.hpp:385]: Please include braces even for single line if/else

File: include/real/exact_number.hpp [hunk 6]
```diff
@@ -240,10 +550,10 @@ namespace boost {
              */
             bool operator>(const exact_number& other) const {
                 std::vector<int> zero = {0};
-                if (this->digits == zero) {
-                    if (other.digits == zero || other.positive)
-                        return false;
-                    else
+                if (this->digits == zero || this->digits.empty()) {
```
Comment (sdavtaker) [inline include/real/exact_number.hpp:555]: When are digits empty?

File: include/real/exact_number.hpp [hunk 7]
```diff
@@ -266,6 +576,14 @@ namespace boost {
                 return other.exponent > this->exponent;
             }
 
+            bool operator>=(const exact_number& other) const {
+                return (((*this) == other) || ((*this) > other));
+            }
+
+            bool operator<=(const exact_number& other) const {
+                return (((*this) == other) || ((*this) < other));
```
Comment (sdavtaker) [inline include/real/exact_number.hpp]: It seems cheaper to check: !((*this) > other)

File: include/real/exact_number.hpp [hunk 8]
```diff
@@ -487,6 +807,13 @@ namespace boost {
                 return this->digits.size();
             }
 
+            bool is_integral() { 
+                if (exponent < 0)
+                    return false;
+                else
+                    /// @TODO: ensure it doesn't overflow
+                    return (exponent_t)digits.size() <= exponent;
```
Comment (sdavtaker) [inline include/real/exact_number.hpp]: I think we keep a space after casting operators.

**Review by sdavtaker — CHANGES_REQUESTED** [review]
Small changes and comments, in general seems great.

**Review by Laouen — APPROVED** [review]

File: include/real/exact_number.hpp (hunk 6 above)
Comment (universenox) [reply include/real/exact_number.hpp:555]: On the default constructor, digits constructs to empty. I suppose instead we could construct to be 0, and get rid of checking for empty

**Comment by sdavtaker** [comment]
If empty has no "meaning", we should not create numbers with empty. If the
meaning is the same as 0, that will be a future source of problems for
sure, having 2 representations of the same is always a nightmare.

File: include/real/exact_number.hpp (hunk 6 above)
Comment (universenox) [reply include/real/exact_number.hpp:555]: I tried changing it but it causes nearly all tests to fail, we can leave that change to later.

**Comment by universenox** [comment]
This has been part of the code since before GSoC19. We can change this in another PR, later, since trying to do it now causes all the tests to fail.

> If empty has no "meaning", we should not create numbers with empty. If the meaning is the same as 0, that will be a future source of problems for sure, having 2 representations of the same is always a nightmare.

**Comment by Laouen** [comment]
I agree with Kim, if that is not part of her changes, then we can add a new issue to address the problem in a future PR but it shouldn't be part of this one.


> This has been part of the code since before GSoC19. We can change this in another PR, later, since trying to do it now causes all the tests to fail.
> 
> > If empty has no "meaning", we should not create numbers with empty. If the meaning is the same as 0, that will be a future source of problems for sure, having 2 representations of the same is always a nightmare.

**Comment by sdavtaker** [comment]
Yeah, sorry forgot to tell. But, not everything I comment in a PR is to
fix. We can just add a ticket or task and go back to it later.

**Comment by universenox** [comment]
Okay then, this should be ready to be merged.

**Review by sdavtaker — APPROVED** [review]

**Pull request merged**